sys/include/barriers.h

```diff
@@ -0,0 +1,298 @@
+ * control registers.
+ *
+ * #### In detail
+ * Accesses to memory that is directly interacting with hardware, such as timer
+ * registers, GPIO registers, or interrupt control registers get easily
+ * optimized out by the compiler. Lets assume an 16-bit timer register is
+ * located at address `0x42` and we want to use that to measure the time a call
+ * to `void do_computing(void)` takes:
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * // BEWARE: This example deliberately contains **BUGS**
+ * #define TIMER    (*((uint16_t *)0x42))
+ *
+ * uint16_t measure_time(void) {
+ *     uint16_t pre = TIMER;
+ *     do_computing();
+ *     uint16_t post = TIMER;
+ *     return pre - post;
+ * }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * In the above example from the compilers perspective the value in `TIMER`
+ * does not change between the two accesses, so the difference is deducted to
+ * always be zero. The compiler, thus, will most likely never access `TIMER`
+ * at all and `measure_time()` will always return `0` instead.
+ *
+ * Memory mapped I/O requires the use of the `volatile` qualifier and is also
+ * the only legitimate use of the `volatile` qualifier: Every access marked as
+ * `volatile` is dutifully performed by the machine code the compiler generates,
+ * even if those accesses seem to make no sense to the compiler. Note the
+ * pitfall that `(*((uint16_t * volatile)0x42))` would be a `volatile` pointer
+ * to regular memory, which is **incorrect** for the scenario at hand. (E.g.
+ * a network controller might indeed update a pointer to a shared buffer, in
+ * which case the pointer itself would have to be `volatile`. But here the timer
+ * register is changed by the hardware, not the pointer to that register.)
+ * The (partially) fixed code would look like this:
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * // BEWARE: This example **still** contains **BUGS**
+ * #define TIMER    (*((volatile uint16_t *)0x42))
+ *
+ * uint16_t measure_time(void) {
+ *     uint16_t pre = TIMER;
+ *     do_computing();
+ *     uint16_t post = TIMER;
+ *     return pre - post;
+ * }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * In the above example the compiler will now dutifully generate code to access
+ * the timer register two times. The compiler will also not reorder any
+ * `volatile` memory access in regard to any other `volatile` memory access.
+ * But some compilers will reorder non-`volatile` accesses across `volatile`
+ * memory accesses. If the implementation of `do_computing()` is visible to the
+ * compiler, the compiler may conclude that do_computing() could be moved
+ * above the first or below the second access to `TIMER`, which will render the
+ * benchmark meaningless. (**Beware:** Even if `do_something()` is an external
```

Comment:
Is this accurate? AFAIK, function calls are sequence points WRT the abstract machine. LTO cannot change that.

Reply:
The compiler can do any transformation that will not yield different observable behavior compared to the abstract machine. If I remember correctly observable behavior is defined as (likely incomplete):

- Side-effects cause by function calls
- Access to memory marked as `volatile`
- The return value of `main()` (at least on non-embedded systems)

If the C compiler can proof that a function does not yield observable behavior, the compiler thus can get as creative as it wants.

Reply:
@kaspar030 _library_ function calls. I made the experiment a while ago with a some functions defined in a single C file (no LTO), mixing accesses to globals and calls. GCC is smart enough to realize that some function do not touch global memory. I'm not sure if LTO is also as smart.

Your point still applies to https://github.com/RIOT-OS/RIOT/blob/11da5e8e671c12e5182c1f7aa651489edbe5390a/core/atomic_c11.c#L354

Reply:
> I made the experiment a while ago with a some functions defined in a single C file (no LTO), mixing accesses to globals and calls. GCC is smart enough to realize that some function do not touch global memory. I'm not sure if LTO is also as smart.

This sounds like anecdotal evidence.

Reply:
@kaspar030: You can observe that compiler barriers do make a difference in https://github.com/RIOT-OS/RIOT/pull/11440, where those are added to `irq_disable()`, `irq_restore()` for the ARMv7 platform etc. Those barriers do not make any difference when compiling without LTO, as the function calls are treated as barriers then. (Unless I missed something in the doc, the ARMv7 platform is in-order, so no hardware memory barriers are used there.)

Reply:
<details>
 <summary>read_fence_test.c</summary>

```c
extern void f(void);

int x;

int *p;

int test(void)
{
    int y = x;
    f();
    return y+x;
}

int z(int x)
{
    return x*x;
}

int test2(void)
{
    int y = x;
    int k = z(x);
    return y+x+k;
}


int test3(void)
{
    int y = x;
    //__sync_synchronize();
    __asm__ __volatile__("": : :"memory");
    return y + x;
}
```

</details>

Compile with `arm-none-eabi-gcc -mcpu=cortex-m3 -std=c99 -Og -dA -ggdb -c read_fence_test.c`

In `test()`, the compiler is forced to emit a re-read of `x` because it does not know if `f()` modified it.

Reply:
> In `test()`, the compiler is forced to emit a re-read of `x` because it does not know if `f()` modified it.

The call to `f()` behaves like a compiler barrier as long as the compiler cannot access the implementation of `f()`. With LTO this might no longer be the case (depending on where `f()` is implemented). But even if `f()` is implemented in a system library, we should not take it for granted to be a compiler barrier for all eternity: When LTO becomes more and more popular, maybe we get LTO-enabled static linking against system libraries (that is, system libraries are not stored in machine code but rather in the internal representation of the compiler to allow optimization over the whole program).

Therefore, I think we should not rely on calls to external function to behave like compiler barriers.